Sources/StreamChat/Database/DatabaseContainer.swift

```diff
@@ -253,13 +253,29 @@ class DatabaseContainer: NSPersistentContainer {
         }
     }
     
-    func read<T>(_ actions: @escaping (DatabaseSession) throws -> T, completion: @escaping (Result<T, Error>) -> Void) {
-        let context = backgroundReadOnlyContext
+    func write(_ actions: @escaping (DatabaseSession) throws -> Void) async throws {
+        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
+            write(actions) { error in
+                if let error {
+                    continuation.resume(throwing: error)
+                } else {
+                    continuation.resume(returning: ())
+                }
+            }
+        }
+    }
+        
+    private func read<T>(
+        from context: NSManagedObjectContext,
+        _ actions: @escaping (DatabaseSession) throws -> T,
+        completion: @escaping (Result<T, Error>) -> Void
+    ) {
         context.perform {
             do {
+                let changeCounts = context.currentChangeCounts()
                 let results = try actions(context)
-                if context.hasChanges {
-                    assertionFailure("Background context is read only, but calling actions() created changes")
+                if changeCounts != context.currentChangeCounts() {
+                    assertionFailure("Context is read only, but actions created changes: (updated=\(context.updatedObjects), inserted=\(context.insertedObjects), deleted=\(context.deletedObjects)")
```

Comment:
Did not figure out why deleted message is kept tracked by the context which triggered this assert. Just changed it around so that we only trigger it when actually the actions closure changed something.